### PR TITLE
Add clink 1.5

### PR DESCRIPTION
**Pull request by clairezc27**
## Change Log ##

- Add a clink to the database upon completion of the Clink modal form

- Validate whether a clink title is unique, display a response message accordingly

## Screenshots ##
Screenshots only available for unique title validation; adding to database functionality can be seen by cloning the repo.
|| Screenshot|
| --- | --- |
| Successful Add | Screenshot 2020-07-13 at 1 47 09 PM - Display 1 |
| Title Exists Error | Screenshot 2020-07-13 at 1 46 51 PM - Display 1 |

## Future Steps ##

- Dynamically display added clinks in Dashboard

- Modify database to account for user mappings to clinks

### CODE REVIEW

File: flask-server/main.py [hunk 1]
```diff
@@ -114,11 +116,46 @@ def decode_auth_token(auth_token):
         return 'Signature expired. Please log in again.'
     except jwt.InvalidTokenError:
         return 'Invalid token. Please log in again.'
+
+@app.route('/apis/add-clink', methods=['POST'])
+def add_clink():
+    try:
```
Comment (michaeleggboy) [inline flask-server/main.py:122]: Could remove try, except statement for consistency for code(sign-up api has the same functionality but w/o try, except statement). I originally added it to provide some kind of error response to test redux store dispatch actions. Realistically there won't be an exception with storing and fetching a clink.

Comment (clairezc27) [reply flask-server/main.py]: done!

File: flask-server/main.py [hunk 2]
```diff
@@ -196,6 +196,32 @@ def decode_auth_token(auth_token):
...
+    titleQuery = datastore_client.query(kind='clink')
+    titleQuery.add_filter('title', '=', request.json['title'])
+    titleResult = list(titleQuery.fetch())
+
+    if titleResult:
+        response_object = {
+            'status': 'fail',
+            'message': 'Title already exists.'
+        }
+        return response_object, 401
```
Comment (marsw42) [inline flask-server/main.py:210]: 401 is for authentication error, use 400 instead.

File: react-app/src/components/Common/newbutton.js [hunk 3]
```diff
@@ -1,158 +1,186 @@
...
+  const [visible, setVisible] = useState(false);
+  const [loading, setLoading] = useState(false);
+  const [form, setForm] = useState('bookmark');
 
-  showModal = () => {
-    this.setState({
-      visible: true,
-    });
+  const showModal = () => {
+    console.log('showmodal')
```
Comment (marsw42) [inline react-app/src/components/Common/newbutton.js:24]: Remove the console.log

File: react-app/src/components/Common/newbutton.js [hunk 4]
```diff
@@ -1,158 +1,186 @@
...
+    setVisible(true);
   };
 
-  onFinish = values => {
-    this.setState({ loading: true })
+  const onBookmarkFinish = values => {
+    console.log(values);
```
Comment (marsw42) [inline react-app/src/components/Common/newbutton.js:29]: Remove console.log

File: react-app/src/components/Common/newbutton.js [hunk 5]
```diff
@@ -1,158 +1,186 @@
...
-    super(props);
-    this.state = { loading: false, visible: false, form: 'bookmark' };
-  }
+  const dispatch = useDispatch()
+  const [bookmarkForm] = Form.useForm()
+  const [clinkForm] = Form.useForm()
+  const key = "formFeedback"
+
+  const [visible, setVisible] = useState(false);
+  const [loading, setLoading] = useState(false);
```
Comment (marsw42) [inline react-app/src/components/Common/newbutton.js:20]: If it's a boolean, I would rename this as setIsLoading and isLoading, same for visible

File: react-app/src/components/Common/newbutton.js [hunk 6]
```diff
@@ -1,158 +1,186 @@
...
     setTimeout(() => {
-      this.setState({ loading: false, visible: false });
+      setLoading(false);
+      setVisible(false);
     }, 3000);
   };
 
-  onFinishFailed = errorInfo => {
+  const onClinkFinish = values => {
+    console.log(values);
```
Comment (marsw42) [inline react-app/src/components/Common/newbutton.js:39]: Remove console.log

File: react-app/src/components/Common/newbutton.js [hunk 7]
```diff
@@ -1,158 +1,186 @@
...
+    dispatch(addClink(values.clinkTitle, addSuccess, addFail))
+    clinkForm.resetFields()
+    setTimeout(() => {
+      setLoading(false);
+      setVisible(false);
+    }, 3000);
+  };
+
+  const addSuccess = () => {
+    console.log("add success")
```
Comment (marsw42) [inline react-app/src/components/Common/newbutton.js:50]: remove console.log

File: react-app/src/components/Common/newbutton.js [hunk 8]
```diff
@@ -1,158 +1,186 @@
...
+    setTimeout(() => {
+      message.success({ content: 'Successfully added clink.', key, duration: 2} );
+    }, 1000)
+  }
+
+
+  const addFail = (error) => {
+    console.log("add fail")
```
Comment (marsw42) [inline react-app/src/components/Common/newbutton.js:58]: remove console.log

File: react-app/src/components/Common/newbutton.js [hunk 9]
```diff
@@ -1,158 +1,186 @@
...
+    setTimeout(() => {
+      message.error({ content: error, key, duration: 2});
+    }, 1000)
+  }
+
+  const onFinishFailed = errorInfo => {
     console.log('Failed:', errorInfo);
```
Comment (marsw42) [inline react-app/src/components/Common/newbutton.js:65]: remove console.log

File: react-app/src/components/Common/newbutton.js [hunk 10]
```diff
@@ -1,158 +1,186 @@
...
+              >
+                <Select mode="multiple" defaultValue="All">
+                  <Option value="All">All</Option>
+                </Select>
+              </Form.Item>
+            </Form>
+          </TabPane>
+          <TabPane tab="Clink" key="clink">
+            <Form
+              {...{ layout: 'vertical' }}
```
Comment (marsw42) [inline react-app/src/components/Common/newbutton.js:157]: just do `layout="vertical"` instead

File: react-app/src/components/Common/newbutton.js [hunk 11]
```diff
@@ -1,158 +1,186 @@
...
-                name="bookmark"
-                initialValues={{
-                  remember: false,
-                }}
-                onFinish={this.onFinish}
-                onFinishFailed={this.onFinishFailed}
+        <Tabs defaultActiveKey={form} onChange={switchForm} >
+          <TabPane tab="Bookmark" key="bookmark">
+            <Form
+              {...{ layout: 'vertical' }}
```
Comment (marsw42) [inline react-app/src/components/Common/newbutton.js:96]: `layout="vertical"`

File: react-app/src/features/clink.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+const initialState = {
+  isAddingClink: false,
+
+};
+
+const clinkSlice = createSlice({
+  name: 'clink',
+  initialState,
+  reducers: {
+    addingClinkStart(state){
```
Comment (marsw42) [inline react-app/src/features/clink.js:13]: `addClinkStart`

**Pull request merged**